src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/IncrementingPollingCounter.cs

```diff
@@ -43,19 +44,40 @@ public IncrementingPollingCounter(string name, EventSource eventSource, Func<dou
         private double _increment;
         private double _prevIncrement;
         private readonly Func<double> _totalValueProvider;
+        private bool _resetStatistics;
+
+        internal void ResetStatistics()
+        {
+            lock (this)
+            {
+                _resetStatistics = true;
+            }
+        }
 
         /// <summary>
         /// Calls "_totalValueProvider" to enqueue the counter value to the queue.
         /// </summary>
-        internal void UpdateMetric()
+        private void UpdateMetric()
         {
-            try
+            lock (this)
             {
-                lock (this)
+                if (_resetStatistics)
```

Comment:
@noahfalk just calling Update twice in a row here doesn't feel right

Reply:
Thank you for review!
While working on this issue I learned that we do reset to fix counter values on first reporting and it seemed okay to me to do 2 calls in a row in this case (https://github.com/dotnet/coreclr/pull/25739).

Reply:
Two calls in a row likely means that all IncrementingPollingCounters would return 0 (or near 0) on the first report because very little time has passed for the counter to be incremented. dotnet/coreclr#25739 was attempting to ensure the first value reported for counters was meaningful. In the case of IncrementingPollingCounter that meant querying the value once, then waiting the duration of one sampling interval and polling again. The reported difference should represent the change in value during one whole sampling time interval.

I think the idea to move the query to the timer thread is the right direction, but we still want it to happen as soon as possible rather than delaying it until the end of the next sampling interval. I think you could accomplish that doing something like this...

Add a new static list:
```C#
class CounterGroup
{
    ...
    private static List<CounterGroup>? s_needsResetCounterGroup;
```

In EnableTimer() add counter groups to that list and don't call ResetStatistics():
```C#
if(!s_needsResetCounterGroup.Contains(this))
{
    s_needsResetCounterGroup.Add(this);
}
// notify the polling thread that the polling interval may have changed and the sleep should
// be recomputed
s_pollingThreadSleepEvent!.Set();
```

In PollForValues() snapshot the list and invoke ResetCounters() on each item in it:
```C#
List<CounterGroup> groupsToReset = new List<CounterGroup>();
while(true)
{
    ...
    lock(s_counterGroupLock)
    {
        if(s_needsResetCounterGroup.Count > 0)
        {
            groupsToReset = s_needsResetCounterGroup;
            s_needsResetCounterGroup = new List<CounterGroup>();
        }
        ...
    }

    foreach(var group in groupsToReset)
    {
        group.ResetStatistics();
    }
    groupsToReset.Clear();
    ...
}
```

And in DisableTimer() remove the counter group if it happens to still be there. This should be a rare case where the timer is enabled and then disabled extremely quickly.
```C#
DisableTimer()
{
    ...
    s_needsResetCounterGroup?.Remove(this);
}
```

I expect in the typical case the polling thread will wake up once because s_pollingThreadSleepEvent was signaled and it will run all the calls to ResetStatistics() at that point. Then at the end of polling interval it will wake up again, observe that polling time has elapsed and start invoking all the EventCounter OnTimer callbacks. Its not perfect as we aren't guaranteed the PollingThread will wake up quickly after the signal, but if the app is so CPU constrained it can't reliably wake up the polling thread on time this won't be the only measurement that gets skewed.

Reply:
Hey, @noahfalk! Thanks a lot for review! Please take a look at my changes

In EnableTimer method I had to add waiting for reset to finish. Without this waiting, tracing tests failed https://dev.azure.com/dnceng-public/public/_build/results?buildId=762610&view=results11. 
I looked at TestIncrementingEventCounter and as far as I understand it failed because after creating SimpleEventSource we immediately increment the counter from the test. At the same time our reset is scheduled to run on Timer thread, its execution will eventually overwrite the values written from the test

I would also add exception handling in PollForValues, to unblock a thread waiting in EnableTimer if there's an exception in Timer thread.

Reply:
Happy to :) I think this current version of the code still has a modified form of the original deadlock and you will need to get rid of blocking on that new event you added to resolve it. I commented a little more detailed inline.